### PR TITLE
nixos/desktop-manager/none: add option to run XDG autostart files

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change

`fcitx5` and `service.earlyoom` rely on use XDG autostart files to start. But for X session with only window manager and no desktop manager (`none` is used), no one can start them.

This options is added to run these autostart files for sessions without desktop manager to make other services just work.

Currently, `fcitx5` has both systemd service and autostart file, which is launched twice if there is a desktop manager (and the second attempt fails), but once if not (eg. using bare `xmonad`). See https://github.com/NixOS/nixpkgs/pull/125303#issuecomment-863121614

@Vonfry for testing.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by Vonfry — APPROVED** [review]
Great work. The service can work.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 127367` at abcf5bf5 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.trivial</li>
  </ul>
</details>

**Comment by oxalica** [comment]
> It seems this has issues with env variables, so you have to do systemctl --user import-environment manually

This also happens on KDE with `systemdBoot=true`. `systemd` is not bash so it's not able to source your `bashrc` or so. Generally you should not rely on external variables inside systemd unit (so do autostart desktop files) and instead explicit you can explicitly write them in your autostart command.
If you really want to set global environment variables, they should go PAM environments (typically `/etc/pam/environment`). They are system-wide but you can refer to HOME or PAM_USER inside. You can set PAM environments through nixos setting `environment.sessionVariables`.

See also https://github.com/systemd/systemd/issues/20049

> Also systemd doesn't seem to generate unit if it has X-GNOME-Autostart-Phase=Initialization X-KDE-autostart-phase=1 so i had to remove those from the desktop entry.

The startup phase is not supported by systemd yet. So explicitly setting them would make them be skipped. For normal non-critical programs, startup phase doesn't really matter and there is no need to specify it yes? There's also an issue about this behavior: https://github.com/systemd/systemd/issues/18791

**Comment by kanashimia** [comment]
> Generally you should not rely on external variables inside systemd unit 

Yeah, but since in `xdg-user-dirs` case the actual configuration is done by placing files using `environment.etc`, you never set variables yourself, it is not _that_ obvious that you rely on env variable to be set, so the option description should note this.

Also users won't even know that this has something to do with systemd units without looking at the source code. So something along the lines of "This uses systemd-xdg-autostart-generator, which creates user service for each desktop entry. Each service name begins with `app-`" 

> The startup phase is not supported by systemd yet. So explicitly setting it would make the desktop file be skipped.

This one should def be noted, _silently_ skipping files is big a problem.

So, basically mention all of this in the description. 
Other than that, this PR is great, i didn't even know that systemd could do this before. (c)

**Comment by oxalica** [comment]
The purpose of this PR is to provide a sane *default* for WM-only users. Currently no autostart files is executed by default. After this PR, they *are* executed by systemd if systemd support them. IMO, there is no "behavior change". If user already uses other program like [dex](https://github.com/jceb/dex) to run their autostart files, they are not affected.

Well, yes, the behavior of systemd *should* be noted somewhere. I'm not sure if it's the best place to put them in the description of `services.xserver.desktopManager.runXdgAutostartIfNone`.

**Comment by ztlevi** [comment]
I'm currently under Gnome, I can verify this works perfectly for me! Thanks

**Review by thiagokokada — APPROVED** [review]
Not sure why I got notified for this PR, but it seems like a good idea and the PR seems simple enough. Also doesn't affect anything and there are even tests.

Is there still interest on this? If so I can merge it.

**Comment by Vonfry** [comment]
I still interest on this. Thanks.

File: nixos/modules/services/x11/desktop-managers/none.nix [hunk 1]
```diff
@@ -1,7 +1,46 @@
...
 {
-  services.xserver.desktopManager.session =
-    [ { name = "none";
-        start = "";
-      }
-    ];
+  options = {
+    services.xserver.desktopManager.runXdgAutostartIfNone = mkOption {
+      type = types.bool;
+      default = true;
```
Comment (thiagokokada) [inline nixos/modules/services/x11/desktop-managers/none.nix:10]: For all people that already uses something else (e.g.: `dex`), having this a default can end up the service being started twice right?

If so, I think it is better to either leave this option as false by default or to have an entry in the release notes. Actually an entry on the release notes is good even if this option is leave as default for false to help discovery.

**Comment by thiagokokada** [comment]
Ok, I just have an issue of having this defaulting to true and no release notes. So either of them should be fixed first (better yet, we could default this to false and have an entry on release notes).

File: nixos/modules/services/x11/desktop-managers/none.nix (hunk 1 above)
Comment (thiagokokada) [reply nixos/modules/services/x11/desktop-managers/none.nix:10]: Also, maybe defaulting to false here is better anyway because I am not sure if people expect to having an auto start manager even when they don't use a a desktop manager.

**Comment by oxalica** [comment]
Rebased and added release notes.

**Comment by thiagokokada** [comment]
@ofborg test xmonad-xdg-autostart

File: nixos/doc/manual/release-notes/rl-2205.section.md [hunk 2]
```diff
@@ -656,6 +656,15 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - The `services.stubby` module was converted to a [settings-style](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) configuration.
 
+- The option `services.xserver.desktopManager.runXdgAutostartIfNone` was added in order to
+  automatically run XDG autostart files for sessions without a desktop manager.
+  This replaces helpers like the `dex` package.
+
+- When setting `i18n.inputMethod.enavled` to `fcitx5`, it no longer creates coresponding systemd user services.
```
Comment (thiagokokada) [inline nixos/doc/manual/release-notes/rl-2205.section.md:663]: Typo: `i18n.inputMethod.enabled`.

Comment (thiagokokada) [reply nixos/doc/manual/release-notes/rl-2205.section.md:663]: Also, could you change all those literal to the link to the option? For example:

```
`i18n.inputMethod.enabled` -> [i18n.inputMethod.enabled](#opt-i18n.inputMethod.enabled)
```

This will make a link to the manual that will allow the users to click on it.

**Review by thiagokokada — COMMENTED** [review]
Some doc changes, otherwise LGTM.

File: nixos/doc/manual/release-notes/rl-2205.section.md (hunk 2 above)
Comment (oxalica) [reply nixos/doc/manual/release-notes/rl-2205.section.md:663]: > Also, could you change all those literal to the link to the option?

Seems it's not enforced in the file. There is only little usage of this syntax.

Comment (thiagokokada) [reply nixos/doc/manual/release-notes/rl-2205.section.md:663]: > Seems it's not enforced in the file. There is only little usage of this syntax.

Still would be better since this makes it easier to switch between the release notes/option description. But your choice.

Anyway, please fix at least the typo.

Comment (oxalica) [reply nixos/doc/manual/release-notes/rl-2205.section.md]: Fixed.

**Comment by thiagokokada** [comment]
@ofborg test xmonad-xdg-autostart

File: nixos/doc/manual/from_md/release-notes/rl-2205.section.xml [hunk 3]
```diff
@@ -1798,6 +1798,29 @@
...
+        </para>
+      </listitem>
+      <listitem>
+        <para>
+          When setting
+          <link linkend="opt-i18n.inputMethod.enabled">i18n.inputMethod.enabled</link>
+          to <literal>fcitx5</literal>, it no longer creates
+          coresponding systemd user services. It now relies on XDG
+          autostart files to start and work properly in your desktop
+          seesions. If your are using only a window manager without a
```
Comment (superherointj) [inline nixos/doc/manual/from_md/release-notes/rl-2205.section.xml:1817]:
```suggestion
          seesions. If your are using only a window manager without a
```
sessions

Comment (superherointj) [inline nixos/doc/manual/from_md/release-notes/rl-2205.section.xml:1817]:
```suggestion
          sessions. If your are using only a window manager without a
```

Comment (thiagokokada) [reply nixos/doc/manual/from_md/release-notes/rl-2205.section.xml:1817]: Sorry @oxalica. Can you also fix those typos?

File: nixos/doc/manual/release-notes/rl-2205.section.md [hunk 4]
```diff
@@ -656,6 +656,17 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - The `services.stubby` module was converted to a [settings-style](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) configuration.
 
+- The option
+  [services.xserver.desktopManager.runXdgAutostartIfNone](#opt-services.xserver.desktopManager.runXdgAutostartIfNone)
+  was added in order to automatically run XDG autostart files for sessions without a desktop manager.
+  This replaces helpers like the `dex` package.
+
+- When setting [i18n.inputMethod.enabled](#opt-i18n.inputMethod.enabled) to `fcitx5`,
+  it no longer creates coresponding systemd user services.
```
Comment (thiagokokada) [inline nixos/doc/manual/release-notes/rl-2205.section.md:665]:
```suggestion
  it no longer creates corresponding systemd user services.
```

**Comment by oxalica** [comment]
@thiagokokada Thank you for the fix. I shouldn't stayed up so long yesterday :cry:
